File: Backend/BridgeMicroservice/utils/error_codes.py
"""Stable, machine-readable codes for every HTTPException Bridge raises.

Why we have this:

  Without structured codes, the frontend either had to keyword-match
  free-form English (brittle) or show the raw English to a Polish /
  Spanish user (ugly). With every failure carrying a `code` field the
  frontend can branch on it, look up a localized string, and the
  English `message` stays as a fallback for codes the UI doesn't know
  yet.

Wire format (since the structured-error refactor):

    HTTPException(detail={"code": "<CODE>", "message": "<msg>"},
                  status_code=<int>)

FastAPI emits this verbatim as `{"detail": {"code": ..., "message": ...}}`
so the frontend reads `detail.code` directly — no string parsing.

Adding a new code:

    1. Add a new constant below in the matching section.
    2. Add a translation key `errors.codes.<CODE>` in
       Frontend/src/config/i18n.ts under all locales.
    3. Use raise_with_code() at the call site.

Don't reuse codes for different conditions — readers grep for them.
"""
from typing import NoReturn

from fastapi import HTTPException


# ── Auth & user context ────────────────────────────────────────────
AUTH_MISSING_USER = "AUTH_MISSING_USER"
AUTH_INVALID_TOKEN = "AUTH_INVALID_TOKEN"

# ── Generic input ──────────────────────────────────────────────────
INPUT_VALIDATION_FAILED = "INPUT_VALIDATION_FAILED"

# ── File / upload ──────────────────────────────────────────────────
FILE_NAME_REQUIRED = "FILE_NAME_REQUIRED"
FILE_TYPE_PDF_ONLY = "FILE_TYPE_PDF_ONLY"
PDF_NO_TEXT = "PDF_NO_TEXT"
PDF_GARBLED_TEXT = "PDF_GARBLED_TEXT"
PDF_AI_REJECTED = "PDF_AI_REJECTED"
FILE_PROCESSING_FAILED = "FILE_PROCESSING_FAILED"

# ── Materials / Quiz ───────────────────────────────────────────────
MATERIALS_NO_RELEVANT = "MATERIALS_NO_RELEVANT"

# ── AI provider (litellm passthrough) ──────────────────────────────
AI_PROVIDER_UNSUPPORTED = "AI_PROVIDER_UNSUPPORTED"
AI_API_KEY_MISSING = "AI_API_KEY_MISSING"
AI_AUTH_FAILED = "AI_AUTH_FAILED"
AI_RATE_LIMITED = "AI_RATE_LIMITED"
AI_TIMEOUT = "AI_TIMEOUT"
AI_BAD_GATEWAY = "AI_BAD_GATEWAY"
AI_EMPTY_RESPONSE = "AI_EMPTY_RESPONSE"
AI_RESPONSE_PARSE_FAILED = "AI_RESPONSE_PARSE_FAILED"

# ── Speaking / Whisper ─────────────────────────────────────────────
SPEAKING_NO_AUDIO = "SPEAKING_NO_AUDIO"
SPEAKING_GROQ_KEY_MISSING = "SPEAKING_GROQ_KEY_MISSING"
SPEAKING_TRANSCRIBE_FAILED = "SPEAKING_TRANSCRIBE_FAILED"
SPEAKING_TRANSCRIBE_PROVIDER_ERROR = "SPEAKING_TRANSCRIBE_PROVIDER_ERROR"
SPEAKING_FEEDBACK_PARSE_FAILED = "SPEAKING_FEEDBACK_PARSE_FAILED"
SPEAKING_FEEDBACK_FAILED = "SPEAKING_FEEDBACK_FAILED"

# ── Task generation ────────────────────────────────────────────────
TASK_INVALID_LEVEL = "TASK_INVALID_LEVEL"
TASK_VALIDATION_FAILED = "TASK_VALIDATION_FAILED"
TASK_GENERATION_FAILED = "TASK_GENERATION_FAILED"

# ── Placement ──────────────────────────────────────────────────────
PLACEMENT_GENERATION_FAILED = "PLACEMENT_GENERATION_FAILED"
PLACEMENT_EVALUATION_FAILED = "PLACEMENT_EVALUATION_FAILED"

# ── User microservice (Bridge -> User) ─────────────────────────────
USER_SERVICE_UNREACHABLE = "USER_SERVICE_UNREACHABLE"
USER_SERVICE_UNAUTHORIZED = "USER_SERVICE_UNAUTHORIZED"
USER_SERVICE_BAD_RESPONSE = "USER_SERVICE_BAD_RESPONSE"
USER_SERVICE_BAD_REQUEST = "USER_SERVICE_BAD_REQUEST"
USER_TOKENS_EMPTY = "USER_TOKENS_EMPTY"


def raise_with_code(
    code: str, status_code: int, message: str
) -> NoReturn:
    """Raise an HTTPException with a structured `{code, message}` body.

    FastAPI surfaces this verbatim as `{"detail": {"code": ...,
    "message": ...}}` — clients read `detail.code` directly, no string
    parsing. The English `message` stays as a fallback for any code
    the frontend hasn't translated yet.
    """
    raise HTTPException(
        status_code=status_code,
        detail={"code": code, "message": message},
    )

File: Backend/UserMicroservice/src/utils/errorCodes.ts
/**
 * Stable, machine-readable codes for every HttpException User raises.
 *
 * Wire contract (since the structured-error refactor):
 *   - throwWithCode() raises an HttpException whose response body is
 *     `{ code, message }`.
 *   - ErrorHandlingMiddleware reads both fields and emits
 *     `payload.code` AND `payload.message` as siblings — code is no
 *     longer embedded inside the message string.
 *   - Frontend `parseApiResponse` reads `payload.code` directly.
 *
 * Adding a new code:
 *   1. Add a constant below in the matching section.
 *   2. Add a translation key `errors.codes.<CODE>` in
 *      Frontend/src/config/i18n.ts under all locales.
 *   3. Use throwWithCode() at the call site.
 */
import {
  BadRequestException,
  ConflictException,
  HttpException,
  HttpStatus,
  InternalServerErrorException,
  NotFoundException,
  UnauthorizedException,
} from "@nestjs/common";

// User-domain codes ────────────────────────────────────────────────
export const USER_ID_REQUIRED = "USER_ID_REQUIRED";
export const USER_NOT_FOUND = "USER_NOT_FOUND";
export const USER_LANGUAGE_ID_MISSING = "USER_LANGUAGE_ID_MISSING";
export const USER_LANGUAGE_NOT_FOUND = "USER_LANGUAGE_NOT_FOUND";
export const USER_LANGUAGE_ALREADY_ADDED = "USER_LANGUAGE_ALREADY_ADDED";
export const USER_INVALID_LEVEL = "USER_INVALID_LEVEL";
export const USER_ID_AND_ROLE_REQUIRED = "USER_ID_AND_ROLE_REQUIRED";
export const USER_AI_TOKEN_CREATE_FAILED = "USER_AI_TOKEN_CREATE_FAILED";
export const USER_AI_TOKEN_NOT_FOUND = "USER_AI_TOKEN_NOT_FOUND";
export const USER_MATERIAL_NOT_FOUND = "USER_MATERIAL_NOT_FOUND";
export const USER_ACHIEVEMENT_NOT_FOUND = "USER_ACHIEVEMENT_NOT_FOUND";
export const USER_HISTORY_MISSING_USER = "USER_HISTORY_MISSING_USER";
export const USER_TOKEN_EXPIRED = "USER_TOKEN_EXPIRED";
export const USER_INVALID_TOKEN = "USER_INVALID_TOKEN";
export const USER_INTERNAL_ERROR = "USER_INTERNAL_ERROR";

function exceptionFor(
  status: number,
  code: string,
  message: string,
): HttpException {
  const body = { code, message };
  switch (status) {
    case HttpStatus.UNAUTHORIZED:
      return new UnauthorizedException(body);
    case HttpStatus.NOT_FOUND:
      return new NotFoundException(body);
    case HttpStatus.CONFLICT:
      return new ConflictException(body);
    case HttpStatus.INTERNAL_SERVER_ERROR:
      return new InternalServerErrorException(body);
    case HttpStatus.BAD_REQUEST:
    default:
      return new BadRequestException(body);
  }
}

/**
 * Throw an HttpException whose response body is `{ code, message }`.
 * The middleware turns that into `payload: { code, message, timestamp }`.
 */
export function throwWithCode(
  code: string,
  status: number,
  message: string,
): never {
  throw exceptionFor(status, code, message);
}
